In [1]:
import pandas as pd

In [2]:
import numpy as np

In [3]:
import matplotlib.pyplot as plt

In [4]:
from sklearn.model_selection import train_test_split

In [5]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [6]:
from sklearn.ensemble import RandomForestClassifier

In [7]:
df = pd.read_csv(r"C:\Users\HP\Desktop\Data Science\Datasets\telecom_copy.csv")

In [24]:
df.info(["Column"])

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 38 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Customer ID                        7043 non-null   object 
 1   Gender                             7043 non-null   object 
 2   Age                                7043 non-null   int64  
 3   Married                            7043 non-null   object 
 4   Number of Dependents               7043 non-null   int64  
 5   City                               7043 non-null   object 
 6   Zip Code                           7043 non-null   int64  
 7   Latitude                           7043 non-null   float64
 8   Longitude                          7043 non-null   float64
 9   Number of Referrals                7043 non-null   int64  
 10  Tenure_in_Months                   7043 non-null   int64  
 11  Offer                              3166 non-null   objec

In [25]:
pd.set_option('display.max_columns', None)
df.head()

,Customer ID,Gender,Age,Married,Number of Dependents,City,Zip Code,Latitude,Longitude,Number of Referrals,Tenure_in_Months,Offer,Phone Service,Avg Monthly Long Distance Charges,Multiple Lines,Internet Service,Internet Type,Avg Monthly GB Download,Online Security,Online Backup,Device Protection Plan,Premium Tech Support,Streaming TV,Streaming Movies,Streaming Music,Unlimited Data,Contract,Paperless Billing,Payment Method,Monthly Charge,Total Charges,Total Refunds,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Customer Status,Churn Category,Churn Reason
0,0002-ORFBO,Female,37,Yes,0,Frazier Park,93225,34.827662,-118.999073,2,9,NaN,Yes,42.39,No,Yes,Cable,16.0,No,Yes,No,Yes,Yes,No,No,Yes,One Year,Yes,Credit Card,65.6,593.30,0.00,0,381.51,974.81,Stayed,NaN,NaN
1,0003-MKNFE,Male,46,No,0,Glendale,91206,34.162515,-118.203869,0,9,NaN,Yes,10.69,Yes,Yes,Cable,10.0,No,No,No,No,No,Yes,Yes,No,Month-to-Month,No,Credit Card,-4.0,542.40,38.33,10,96.21,610.28,Stayed,NaN,NaN
2,0004-TLHLJ,Male,50,No,0,Costa Mesa,92627,33.645672,-117.922613,0,4,Offer E,Yes,33.65,No,Yes,Fiber Optic,30.0,No,No,Yes,No,No,No,No,Yes,Month-to-Month,Yes,Bank Withdrawal,73.9,280.85,0.00,0,134.60,415.45,Churned,Competitor,Competitor had better devices
3,0011-IGKFF,Male,78,Yes,0,Martinez,94553,38.014457,-122.115432,1,13,Offer D,Yes,27.82,No,Yes,Fiber Optic,4.0,No,Yes,Yes,No,Yes,Yes,No,Yes,Month-to-Month,Yes,Bank Withdrawal,98.0,1237.85,0.00,0,361.66,1599.51,Churned,Dissatisfaction,Product dissatisfaction
4,0013-EXCHZ,Female,75,Yes,0,Camarillo,93010,34.227846,-119.079903,3,3,NaN,Yes,7.38,No,Yes,Fiber Optic,11.0,No,No,No,Yes,Yes,No,No,Yes,Month-to-Month,Yes,Credit Card,83.9,267.40,0.00,0,22.14,289.54,Churned,Dissatisfaction,Network reliability


<h5>Call All the column to understand the data type <h5>

In [30]:
df = df.astype({col: "string" for col in df.select_dtypes(include="object").columns})

In [32]:
df.dtypes

Customer ID                          string[python]
Gender                               string[python]
Age                                           int64
Married                              string[python]
Number of Dependents                          int64
City                                 string[python]
Zip Code                                      int64
Latitude                                    float64
Longitude                                   float64
Number of Referrals                           int64
Tenure_in_Months                              int64
Offer                                string[python]
Phone Service                        string[python]
Avg Monthly Long Distance Charges           float64
Multiple Lines                       string[python]
Internet Service                     string[python]
Internet Type                        string[python]
Avg Monthly GB Download                     float64
Online Security                      string[python]
Online Backu

In [37]:
### Change the onject type in to String

In [34]:
df.shape

(7043, 38)

In [35]:
df.describe()

,Age,Number of Dependents,Zip Code,Latitude,Longitude,Number of Referrals,Tenure_in_Months,Avg Monthly Long Distance Charges,Avg Monthly GB Download,Monthly Charge,Total Charges,Total Refunds,Total Extra Data Charges,Total Long Distance Charges,Total Revenue
count,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,6361.000000,5517.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000
mean,46.509726,0.468692,93486.070567,36.197455,-119.756684,1.951867,32.386767,25.420517,26.189958,63.596131,2280.381264,1.962182,6.860713,749.099262,3034.379056
std,16.750352,0.962802,1856.767505,2.468929,2.154425,3.001199,24.542061,14.200374,19.586585,31.204743,2266.220462,7.902614,25.104978,846.660055,2865.204542
min,19.000000,0.000000,90001.000000,32.555828,-124.301372,0.000000,1.000000,1.010000,2.000000,-10.000000,18.800000,0.000000,0.000000,0.000000,21.360000
25%,32.000000,0.000000,92101.000000,33.990646,-121.788090,0.000000,9.000000,13.050000,13.000000,30.400000,400.150000,0.000000,0.000000,70.545000,605.610000
50%,46.000000,0.000000,93518.000000,36.205465,-119.595293,0.000000,29.000000,25.690000,21.000000,70.050000,1394.550000,0.000000,0.000000,401.440000,2108.640000
75%,60.000000,0.000000,95329.000000,38.161321,-117.969795,3.000000,55.000000,37.680000,30.000000,89.750000,3786.600000,0.000000,0.000000,1191.100000,4801.145000
max,80.000000,9.000000,96150.000000,41.962127,-114.192901,11.000000,72.000000,49.990000,85.000000,118.750000,8684.800000,49.790000,150.000000,3564.720000,11979.340000


# Here we undertand the data as The average Monthly charge is 63.596 

In [16]:
df.isnull().sum()

Customer ID                             0
Gender                                  0
Age                                     0
Married                                 0
Number of Dependents                    0
City                                    0
Zip Code                                0
Latitude                                0
Longitude                               0
Number of Referrals                     0
Tenure_in_Months                        0
Offer                                3877
Phone Service                           0
Avg Monthly Long Distance Charges     682
Multiple Lines                        682
Internet Service                        0
Internet Type                        1526
Avg Monthly GB Download              1526
Online Security                      1526
Online Backup                        1526
Device Protection Plan               1526
Premium Tech Support                 1526
Streaming TV                         1526
Streaming Movies                  

In [21]:
df['Offer'] = df['Offer'].fillna('No Offer')

In [23]:
df = df.replace("None", "No")

In [8]:
df.isnull().sum()

Customer ID                          0
Gender                               0
Age                                  0
Married                              0
Number of Dependents                 0
City                                 0
Zip Code                             0
Latitude                             0
Longitude                            0
Number of Referrals                  0
Tenure_in_Months                     0
Offer                                0
Phone Service                        0
Avg Monthly Long Distance Charges    0
Multiple Lines                       0
Internet Service                     0
Internet Type                        0
Avg Monthly GB Download              0
Online Security                      0
Online Backup                        0
Device Protection Plan               0
Premium Tech Support                 0
Streaming TV                         0
Streaming Movies                     0
Streaming Music                      0
Unlimited Data           

In [9]:
df.groupby('Age')['Total Revenue'].mean()

Age
19    2777.119921
20    2962.576693
21    3240.065143
22    3153.365077
23    2799.218562
         ...     
76    3411.738406
77    3847.724444
78    3308.526190
79    3262.208816
80    3513.573030
Name: Total Revenue, Length: 62, dtype: float64

In [10]:
df.groupby('City')['Total Revenue'].sum().sort_values(ascending=False)

City
Los Angeles      852725.23
San Diego        738416.01
Sacramento       353371.84
San Jose         326478.36
San Francisco    306995.99
                   ...    
Highland           1193.64
Loleta              873.05
Angelus Oaks        744.72
Dana Point          694.86
Eagleville          238.57
Name: Total Revenue, Length: 1106, dtype: float64

In [11]:
df.groupby('Tenure_in_Months')['Total Revenue'].mean()

Tenure_in_Months
1       75.365318
2      160.533109
3      243.475350
4      323.940511
5      419.792180
         ...     
68    6480.941700
69    6523.348000
70    6880.454706
71    6871.524353
72    7445.366713
Name: Total Revenue, Length: 72, dtype: float64

In [12]:
df.groupby('Offer')['Total Revenue'].mean()


Offer
No Offer    2926.847702
Offer A     7029.269558
Offer B     5082.423265
Offer C     2759.245470
Offer D     1315.726312
Offer E      302.424969
Name: Total Revenue, dtype: float64

In [13]:
df[['Monthly Charge','Total Charges','Total Revenue']].describe()

,Monthly Charge,Total Charges,Total Revenue
count,7043.000000,7043.000000,7043.000000
mean,63.596131,2280.381264,3034.379056
std,31.204743,2266.220462,2865.204542
min,-10.000000,18.800000,21.360000
25%,30.400000,400.150000,605.610000
50%,70.050000,1394.550000,2108.640000
75%,89.750000,3786.600000,4801.145000
max,118.750000,8684.800000,11979.340000


In [14]:
df.groupby('Customer Status')[['Total Refunds','Total Extra Data Charges']].sum()

,Total Refunds,Total Extra Data Charges
Customer Status,,
Churned,2838.98,13380
Joined,108.35,1080
Stayed,10872.32,33860


In [15]:
df['Profit'] = df['Total Revenue'] - df['Total Refunds']
df.groupby('Customer Status')['Profit'].mean()

Customer Status
Churned    1969.834585
Joined      119.320264
Stayed     3733.372839
Name: Profit, dtype: float64

In [16]:
df.groupby('Contract')['Profit'].mean()

Contract
Month-to-Month    1705.464440
One Year          3979.679987
Two Year          4796.644413
Name: Profit, dtype: float64

In [17]:
# Create a Profit column
df['Profit'] = df['Total Revenue'] - df['Total Refunds']

In [18]:
df.groupby('Customer Status').agg({
    'Total Revenue': 'mean',
    'Total Refunds': 'mean',
    'Profit': ['mean','sum']
})

Total Revenue Total Refunds       Profit             
                         mean          mean         mean          sum
Customer Status                                                      
Churned           1971.353569      1.518983  1969.834585   3681620.84
Joined             119.558921      0.238656   119.320264     54171.40
Stayed            3735.676297      2.303458  3733.372839  17621519.80

In [19]:
df.groupby('Payment Method').agg({
    'Profit': 'mean',
    'Total Refunds': 'sum'
})

,Profit,Total Refunds
Payment Method,,
Bank Withdrawal,3239.593333,8520.96
Credit Card,2965.565431,4846.79
Mailed Check,1406.239818,451.90


In [20]:
prob_profitable = (df['Profit'] > 0).mean()
print(prob_profitable)

1.0


In [21]:
prob_offerA_profit = (df[df['Offer'] == 'Offer A']['Profit'] > 0).mean()

In [22]:
prob_offerA_profit

np.float64(1.0)

In [23]:
df =df.drop(columns=["Customer ID", "Zip Code", "Latitude", "Longitude"])

In [24]:
df['Customer Status'].value_counts(normalize=True)

Customer Status
Stayed     0.670169
Churned    0.265370
Joined     0.064461
Name: proportion, dtype: float64

## This confirms churn is a meaningful problem — enough data to model.

In [25]:
df.groupby('Contract')['Customer Status'].value_counts(normalize=True)

Contract        Customer Status
Month-to-Month  Churned            0.458449
                Stayed             0.428532
                Joined             0.113019
One Year        Stayed             0.877419
                Churned            0.107097
                Joined             0.015484
Two Year        Stayed             0.962825
                Churned            0.025491
                Joined             0.011683
Name: proportion, dtype: float64

## Contract type is a strong churn predictor.

In [26]:
df.groupby('Customer Status')['Monthly Charge'].mean()

Customer Status
Churned    73.347592
Joined     42.775991
Stayed     61.737415
Name: Monthly Charge, dtype: float64

##  MonthlyCharge is a strong churn driver — high charges increase churn.

In [27]:
df.groupby('Tenure_in_Months')['Customer Status'].value_counts(normalize=True).head(10)

Tenure_in_Months  Customer Status
1                 Churned            0.619902
                  Joined             0.380098
2                 Churned            0.516807
                  Joined             0.483193
3                 Joined             0.530000
                  Churned            0.470000
4                 Stayed             0.528409
                  Churned            0.471591
5                 Stayed             0.518797
                  Churned            0.481203
Name: proportion, dtype: float64

In [28]:
df.groupby('Offer')['Customer Status'].value_counts(normalize=True)

Offer     Customer Status
No Offer  Stayed             0.656951
          Churned            0.271086
          Joined             0.071963
Offer A   Stayed             0.932692
          Churned            0.067308
Offer B   Stayed             0.877427
          Churned            0.122573
Offer C   Stayed             0.771084
          Churned            0.228916
Offer D   Stayed             0.732558
          Churned            0.267442
Offer E   Churned            0.529193
          Stayed             0.253416
          Joined             0.217391
Name: proportion, dtype: float64

## Offer type is critical — some offers reduce churn, others increase it.

In [29]:
df['Profit'] = df["Total Revenue"] - df['Total Refunds']

In [30]:
df.groupby('Customer Status')['Profit'].mean()

Customer Status
Churned    1969.834585
Joined      119.320264
Stayed     3733.372839
Name: Profit, dtype: float64

## Profit is directly tied to churn — churned customers generate less profit.

 ## {Contract, Tenure, MonthlyCharge, Offer, PaymentMethod, InternetService, Profit}

In [31]:
pd.get_dummies(df['Contract'], prefix='Contract')

,Contract_Month-to-Month,Contract_One Year,Contract_Two Year
0,False,True,False
1,True,False,False
2,True,False,False
3,True,False,False
4,True,False,False
...,...,...,...
7038,False,True,False
7039,True,False,False
7040,True,False,False
7041,False,False,True


In [32]:
pd.get_dummies(df['Offer'], prefix='Offer')

,Offer_No Offer,Offer_Offer A,Offer_Offer B,Offer_Offer C,Offer_Offer D,Offer_Offer E
0,True,False,False,False,False,False
1,True,False,False,False,False,False
2,False,False,False,False,False,True
3,False,False,False,False,True,False
4,True,False,False,False,False,False
...,...,...,...,...,...,...
7038,False,False,False,False,True,False
7039,False,False,False,False,True,False
7040,False,False,False,False,False,True
7041,False,True,False,False,False,False


In [33]:
pd.get_dummies(df['Payment Method'], prefix='Payment')

,Payment_Bank Withdrawal,Payment_Credit Card,Payment_Mailed Check
0,False,True,False
1,False,True,False
2,True,False,False
3,True,False,False
4,False,True,False
...,...,...,...
7038,False,True,False
7039,True,False,False
7040,False,True,False
7041,False,True,False


In [34]:
pd.get_dummies(df['Internet Service'], prefix='Internet')

,Internet_Yes
0,True
1,True
2,True
3,True
4,True
...,...
7038,True
7039,True
7040,True
7041,True


In [35]:
features = ['Contract','Tenure_in_Months','Monthly Charge',
            'Offer','Payment Method','Internet Service','Profit']

X = df[features]

In [36]:
df['Customer Status'] = df['Customer Status'].replace({'Joined':'Stayed'})

y = df['Customer Status'].map({'Stayed':0, 'Churned':1})

In [37]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 33 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Gender                             7043 non-null   object 
 1   Age                                7043 non-null   int64  
 2   Married                            7043 non-null   object 
 3   Number of Dependents               7043 non-null   int64  
 4   City                               7043 non-null   object 
 5   Number of Referrals                7043 non-null   int64  
 6   Tenure_in_Months                   7043 non-null   int64  
 7   Offer                              7043 non-null   object 
 8   Phone Service                      7043 non-null   object 
 9   Avg Monthly Long Distance Charges  7043 non-null   object 
 10  Multiple Lines                     7043 non-null   object 
 11  Internet Service                   7043 non-null   objec

In [38]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

In [42]:
rf = RandomForestClassifier(
    n_estimators=200,
    random_state=42,
    class_weight='balanced'
)
rf.fit(X_train, y_train)

ValueError: could not convert string to float: 'Month-to-Month'

In [41]:
# Select features
features = ['Contract','Tenure_in_Months','Monthly Charge',
            'Offer','Payment Method','Internet Service','Profit']

X = df[features]

# One-hot encode categorical features
X = pd.get_dummies(X, 
                   columns=['Contract','Offer','Payment Method','Internet Service'], 
                   drop_first=True)

# Target column (Joined → Stayed, then binary map)
df['Customer Status'] = df['Customer Status'].replace({'Joined':'Stayed'})
y = df['Customer Status'].map({'Stayed':0, 'Churned':1})
